bpf/k_tracer.c

```diff
@@ -674,17 +701,28 @@ int app_ingress(struct __sk_buff *skb) {
         return 0;
     }
 
-    unsigned char buf[12];
+    if (!tcp_syn(&tcp) || tcp_ack(&tcp)) {
+        return 0;
+    }
 
-    if (tcp_syn(&tcp) && !tcp_ack(&tcp)) {
-        bpf_skb_load_bytes(skb, tcp.hdr_len, &buf, 4);
+    if (skb->len - tcp.hdr_len < sizeof(tp_info_pid_t)) {
+        bpf_printk("SYN packet without tp info");
+        return 0;
+    }
 
-        s32 len = skb->len-sizeof(u32);
-        bpf_printk("SYN packed len = %d, offset = %d, hdr_len %d", skb->len, len, tcp.hdr_len);
+    tp_info_pid_t tp = {0};
+
+    bpf_skb_load_bytes(skb, tcp.hdr_len, &tp, sizeof(tp_info_pid_t));
+
+    s32 len = skb->len-sizeof(u32);
+    bpf_printk("Received SYN packed len = %d, offset = %d, hdr_len %d", skb->len, len, tcp.hdr_len);
+
+    unsigned char tp_buf[TP_MAX_VAL_LENGTH];
+    make_tp_string(tp_buf, &tp.tp);
+    bpf_printk("tp: %s", tp_buf);
+
+    bpf_map_update_elem(&incoming_trace_map, &conn, &tp, BPF_ANY);
```

Comment:
Once we receive a traceID over the wire (TCP packet) we store it for later to be used by the trace code.